### PR TITLE
support generator style extensions hooks.

### LINKED ISSUE #1864: Extension: Use context managers for lifecycle hooks
## Feature Request Type

- [ ] Core functionality
- [x] Alteration (enhancement/optimization) of existing feature(s)
- [ ] New behavior

## Description

Introduce a new interface for extension hooks - on request/validation/parsing/execution using context managers instead of on start/end hooks:
```py
def on_request(self) -> Iterable[None]:
    #  This part is called when a GraphQL request starts
    yield
    # This part is called when a GraphQL request ends
```
This makes it easier to manage resources (e.g. open connections) and share data between start/end hooks

### PR DESCRIPTION
**Pull request by nrbnlulu**
## Description
Support generator styled Extensions.
i.e:
```python
def on_execute(self):
    #  This part is called before the execution starts
    yield
    #  This part is called after the execution ends.
```

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR
closes #1864 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Nir for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/extensions/base_extension.py
Comment (nrbnlulu) [inline strawberry/extensions/base_extension.py:11]: This is renamed from `on_request` to `on_operation` since one request can possibly run many graphlq operations and this hook would be called many times.

File: strawberry/extensions/context.py
Comment (nrbnlulu) [inline strawberry/extensions/context.py:86]: I have no idea why codecov couldn't catch these lines though they are most definitely got hitted.

**Comment by nrbnlulu** [comment]
@patrick91 for some reason codecov fails the CI, any idea?

**Comment by patrick91** [comment]
@nrbnlulu not sure, I had this earlier as well, maybe another push later should fix it

**Comment by skilkis** [comment]
@nrbnlulu I discussed this with @patrick91 and we both agree that it would be better to have an `Extension` class with hooks defined on it instead having an `_ExtensionHinter`. I've benchmarked the `_ExtensionHinter` approach and several others. Yours performs the best, followed very closely by a metaclass-flavored approach. However, due to readability I suggest going for the latter.

```bash
attribute lookup (hinter) miss  0.0699372
attribute lookup (hinter) hit   0.2921049
meta method miss                0.08605809999999997
meta method hit                 0.2871588
always call method              0.2216319
method comparison miss          0.11121799999999993
method comparison hit           0.2967436000000001
patched method miss             0.1564654000000001
patched method hit              0.47967289999999996
```

That approach registers overridden hooks of an extension at class creation time and then performs a look-up at runtime to retrieve the overridden method:

```python
from __future__ import annotations

from typing import Callable, Dict, Generator, Tuple


class Extension:
    """Utilize class dictionary to determine overrides at class creation time."""

    __hooks__: Dict[Hook, Hook]

    def __init_subclass__(cls) -> None:
        """Register overridden hooks."""
        # Mapping of base hook to overridden hook
        hooks_map: Dict[Callable, Callable] = getattr(cls, "__hooks__", {})

        # Prevent side-effecting super class
        overridden_hooks_map: Dict[Callable, Callable] = {}
        for hook in BASE_HOOKS:
            if hook.__name__ in cls.__dict__:
                overridden_hooks_map[hook] = getattr(cls, hook.__name__)

        # Merge overrides with super class
        setattr(cls, "__hooks__", {**hooks_map, **overridden_hooks_map})

    def on_operation(self):
        yield


Hook = Callable[[Extension], Generator[None, None, None]]

BASE_HOOKS: Tuple[Hook] = (Extension.on_operation,)  # All hooks go here


class RequestContextManager:

    hook: Hook = Extension.on_operation
    extension: Extension = Extension()

    def run_hook(self):
        ext = self.extension
        hook = ext.__hooks__.get(self.hook, None)
        if hook:
            next(hook(ext))
```

<details>
  <summary>Benchmark Code</summary>
  
  ```python
"""Test performance of various ways to determine if a method has been overridden."""

from __future__ import annotations

from timeit import timeit
from typing import Callable, Dict, Generator, Tuple

N = 1000000


class Extension:
    """Base extension, also used to test approach where the method is always called."""

    def on_operation(self):
        yield


class PatchedExtension:
    """Used to test approach where a attribute is patched into base methods."""

    def on_operation(self):
        yield

    on_operation.default = True  # type: ignore


class PatchedExtensionWithMethod(PatchedExtension):
    """Derived method will not have a :py:attr:`default`."""

    def on_operation(self):
        yield


class PatchedExtensionWithNoMethod(PatchedExtension):
    pass


class ExtensionWithNoMethod:
    """Equivalent to the _ExtensionHinter approach."""

    pass


class DerivedExtensionWithNoMethod(Extension):
    """Inheriting without specializing."""

    pass


class DerivedExtensionWithMethod(Extension):
    """Inheriting and specializing. This requires checking if method is different."""

    def on_operation(self):
        yield


class MetaExtension:
    """Utilize class dictionary to determine overrides at class creation time."""

    __hooks__: Dict[Hook, Hook]

    def __init_subclass__(cls) -> None:
        hooks_map: Dict[Callable, Callable] = getattr(cls, "__hooks__", {})

        # Prevent side-effecting super class
        overridden_hooks_map: Dict[Callable, Callable] = {}
        for hook in BASE_HOOKS:
            if hook.__name__ in cls.__dict__:
                overridden_hooks_map[hook] = getattr(cls, hook.__name__)

        # Merge overrides with super class
        setattr(cls, "__hooks__", {**hooks_map, **overridden_hooks_map})

    def on_operation(self):
        yield


Hook = Callable[[MetaExtension], Generator[None, None, None]]

BASE_HOOKS: Tuple[Hook] = (MetaExtension.on_operation,)


class DerivedMetaExtensionWithMethod(MetaExtension):
    def on_operation(self):
        yield


class DerivedMetaExtensionWithNoMethod(MetaExtension):
    pass


ext_a = Extension()
ext_b = ExtensionWithNoMethod()
ext_c = DerivedExtensionWithNoMethod()
ext_d = DerivedExtensionWithMethod()
ext_e = PatchedExtensionWithNoMethod()
ext_f = PatchedExtensionWithMethod()
ext_g = DerivedMetaExtensionWithNoMethod()
ext_h = DerivedMetaExtensionWithMethod()
base_meta_hook = MetaExtension.on_operation
base_method = Extension.on_operation
patched_base_method = PatchedExtension.on_operation


def test_method_call():
    next(ext_a.on_operation())


def test_hinter_miss():
    method = getattr(ext_b, "on_operation", None)
    if method:
        next(method())


def test_hinter_hit():
    method = getattr(ext_a, "on_operation", None)
    if method:
        next(method())


def test_method_comparison_miss():
    method = getattr(ext_c, "on_operation")
    if method.__func__ is not base_method:
        next(method())


def test_method_comparison_hit():
    method = getattr(ext_d, "on_operation")
    if method.__func__ is not base_method:
        next(method())


def test_patch_miss():
    method = getattr(ext_e, "on_operation")
    if not hasattr(method, "default"):
        next(method())


def test_patch_hit():
    method = getattr(ext_f, "on_operation")
    if not hasattr(method, "default"):
        next(method())

def test_meta_miss():
    hook = ext_g.__hooks__.get(base_meta_hook, None)
    if hook:
        next(hook(ext_g))

def test_meta_hit():
    hook = ext_h.__hooks__.get(base_meta_hook, None)
    if hook:
        next(hook(ext_h))


print("attribute lookup (hinter) miss ", timeit(test_hinter_miss, number=N))
print("attribute lookup (hinter) hit  ", timeit(test_hinter_hit, number=N))
print("meta method miss               ", timeit(test_meta_miss, number=N))
print("meta method hit                ", timeit(test_meta_hit, number=N))
print("always call method             ", timeit(test_method_call, number=N))
print("method comparison miss         ", timeit(test_method_comparison_miss, number=N))
print("method comparison hit          ", timeit(test_method_comparison_hit, number=N))
print("patched method miss            ", timeit(test_patch_miss, number=N))
print("patched method hit             ", timeit(test_patch_hit, number=N))

  ```
</details>

**Comment by nrbnlulu** [comment]
> due to readability .
@skilkis IMO your `__hooks__` is more magical than my `_ExtensionHinter` hack... though. Also does your approuch support inheritance?

**Comment by skilkis** [comment]
@nrbnlulu just the ones that are overridden by the user. This suggestion is to change how the detection of the override works so that we can have a `Extension` class with methods defined on it, instead of needing the `_ExtensionHinter` approach with an empty `Extension` class as currently the override is detected based on the presence of a method (i.e. if method is defined -> run, if method is None -> skip)

**Comment by nrbnlulu** [comment]
@skilkis So which of your extensions approuch? 
Also I think your benchmark is misleading since you don't need to do all of these check at (extension) execution time,
you can cache it somewhere and just call what cached.

**Comment by skilkis** [comment]
@nrbnlulu the `__init_subclass__` approach which I posted above. As for the benchmark, the `ExtensionRunner` will instantiate `ExtensionContextManagerBase` derived classes on every request and since from what I could understand you checking for which hooks are defined during initialization. Therefore, you're right this check doesn't happen while executing the hook, but it must happen at runtime for each `Extension`. The benchmark is a highly simplified version of what I understood from this new implementation.

**Comment by nrbnlulu** [comment]
@skilkis I took a simpler approuch ig, I just check if the method was defined on another file :smile: .

**Comment by nrbnlulu** [comment]
I did my best, the execution order thing made things really complicated so this is why context.py is not very readable
So if you have any idea's how to implement it better while supporting
- execution order 
- initialize generator multiple times (for subscriptions that a requriement) 
I'm happy to hear.

**Comment by nrbnlulu** [comment]
It is starting to be a burden to maintain this PR, is this PR wanted? @patrick91

**Comment by patrick91** [comment]
> It is starting to be a burden to maintain this PR, is this PR wanted? @patrick91

@nrbnlulu yes, sorry :) @skilkis shall we do a call on discord this week about this PR?

**Comment by nrbnlulu** [comment]
OK. Just a reminder that this PR and the #2430 are closely related.

**Comment by skilkis** [comment]
@nrbnlulu thanks for maintaining this PR and your patience, it is definitely wanted! As we discussed on Discord I will push my proposed change to this branch, then it should be good to go pending the approval of @patrick91 🎉

**Comment by skilkis** [comment]
> > It is starting to be a burden to maintain this PR, is this PR wanted? @patrick91
> 
> @nrbnlulu yes, sorry :) @skilkis shall we do a call on discord this week about this PR?

Definitely, can do it tonight if you have the time, I'm a lot more flexible this week, so anytime works for me 😊

**Review by patrick91 — APPROVED** [review]
Just did a final review with @skilkis, we pushed a couple of minor fixes (just for typing).

Thanks for this amazing PR @nrbnlulu!

**Pull request merged**
